Don't include_docs if you just want the case ID

kaapstorm opened the pull request:
@snopoke, like so?

(The diff looks really ugly if **Unified**. More readable with **Split** or **View**.)

cc @benrudolph

snopoke commented on corehq/apps/hqcase/utils.py:
```diff
@@ -71,22 +71,35 @@ def pact_wrapper(data):
+    """
+    rows = _get_cases_by_domain_hq_user_id(domain, user_id, include_docs=False)
+    for row in rows:
+        if row['type'] == case_type:
```
Ah damn, I forgot about the type filtering. If you look at the view I think you'll find that there is no `type` in the row. One option would be to output the case type in the view `value` though that would require a reindex.

snopoke replied:
Looking at the usage of `get_cases_by_domain_hq_user_id` it is always queried with a case type so I think it would be better to add the case type to the view key:
```javascript
emit([doc.domain, doc.hq_user_id, doc.type], null);
```

kaapstorm committed:
Use include_docs=False to return only case ID

dimagimon commented on corehq/apps/hqcase/tests/test_get_case.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class GetCaseTest(TestCase):
+    
```
W293 blank line contains whitespace

kaapstorm commented:
Refactored. View key includes case type. Calls changed to use appropriate method.

I also added a little test just to make sure I hadn't missed anything. Not sure if it's at all useful. Happy to drop that commit if the test doesn't actually tell us anything that other tests aren't going to tell us anyway.

benrudolph commented on corehq/apps/hqcase/_design/views/by_domain_hq_user_id/map.js:
```diff
@@ -1,5 +1,5 @@
 function(doc){
     if(doc.doc_type == "CommCareCase" && doc.hq_user_id) {
-        emit([doc.domain, doc.hq_user_id], null);
+        emit([doc.domain, doc.hq_user_id, doc.type], null);
```
Now that we have to re index we may want to move this to a different design doc. This way in the future if we modify the view it doesn't cause all other views in the design doc to reindex. This was one of the takeaways from talking with cloudant

kaapstorm replied:
That's good to know! I've created a new "by_domain_hq_user_id_type". And I've just realised that `get_callcenter_case_mapping()` uses `by_domain_hq_user_id` with just `domain` and `hq_user_id`, so this seems like the way to go.

benrudolph commented:
Nice one. Good tests, thanks for following up

dimagimon commented on corehq/apps/hqcase/_design/views/by_domain_hq_user_id_type/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc){
```
Missing name in function declaration.

kaapstorm commented:
:white_check_mark:

benrudolph commented:
@kaapstorm i've been holding off on merge since i'm not sure when we want to run the reindex. @dannyroberts do you have anything else planned for today or do you think this is ok to merge?

kaapstorm commented:
No rush, @benrudolph -- and before we merge, I created the new `by_domain_hq_user_id_type/{reduce,map}.js` files in the wrong place. I'll move them.

dimagimon commented on corehq/couchapps/cases_by_owner/views/by_domain_hq_user_id_type/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc){
```
Missing name in function declaration.

snopoke commented:
Yes, that's the right place. The only thing I'd change is to go with the naming convention that just got agreed to (https://github.com/dimagi/commcare-hq/pull/8383#issuecomment-157071364) so something like `case_by_domain_hq_user_id_type/view`. Here's another example: https://github.com/dimagi/commcare-hq/pull/9137/files

dimagimon commented on corehq/couchapps/case_by_domain_hq_user_id_type/views/by_domain_hq_user_id_type/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc){
```
Missing name in function declaration.

kaapstorm commented:
:+1: Thanks Simon.

kaapstorm committed:
Moved design doc to one-view convention

kaapstorm committed:
Added test

dimagimon commented on corehq/couchapps/case_by_domain_hq_user_id_type/views/view/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc){
```
Missing name in function declaration.

snopoke commented on corehq/apps/hqcase/utils.py:
```diff
@@ -71,20 +71,29 @@ def pact_wrapper(data):
-        include_docs=True
-    ).all()
+    cases = _get_cases_by_domain_hq_user_id(domain, user_id, case_type, include_docs=True)
+    return cases[0] if cases else None
```
Should these raise an exception if multiple results are returned?

snopoke commented:
This is GTG from me. Just need to schedule the reindex. @dannyroberts we almost need a place to keep track of all these upcoming re-indexes.

dannyroberts commented:
I just keep a link around to https://github.com/dimagi/commcare-hq/pulls?q=is%3Aopen+is%3Apr+label%3Areindex%2Fmigration (as a menubar bookmark)

dannyroberts commented:
And https://github.com/dimagi/commcare-hq/pull/9012 is next up for this Friday. Is this PR ok to wait for an extra week?

kaapstorm commented:
Fine by me.

dannyroberts commented:
I'm going to merge this and start the reindex

@kaapstorm can you also remove other usages of the old view so that we can eventually remove it (will require reindexing _design/hqcase due to the point about reindexing under https://gist.github.com/dannyroberts/8d514fb6460a9f4f0404#view)? From the looks of it, it's only used in `get_callcenter_case_mapping` and that function's not called anywhere, so if you confirm that's the case, it might mean just removing that function.

kaapstorm commented:
Affirmative @dannyroberts. See #9434.
